src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,416 @@
+          break;
+        }
+        case AggregateFunction::Count: {
+          using AggregateType = typename AggregateTraits<ColumnDataType, AggregateFunction::Count>::AggregateType;
+          auto aggregate_function = AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::Count>()
+                                        .get_aggregate_function();
+          _aggregate_values<ColumnDataType, AggregateType, AggregateFunction::Count>(
+              aggregate_group_offsets, aggregate_index, aggregate_function, sorted_table);
+          break;
+        }
+        case AggregateFunction::CountDistinct: {
+          using AggregateType =
+              typename AggregateTraits<ColumnDataType, AggregateFunction::CountDistinct>::AggregateType;
+          auto aggregate_function =
+              AggregateFunctionBuilder<ColumnDataType, AggregateType, AggregateFunction::CountDistinct>()
+                  .get_aggregate_function();
+          _aggregate_values<ColumnDataType, AggregateType, AggregateFunction::CountDistinct>(
+              aggregate_group_offsets, aggregate_index, aggregate_function, sorted_table);
+          break;
+        }
+      }
+    });
+
+    aggregate_index++;
+  }
+
+  // append output to result table
+  result_table->append_chunk(_output_segments);
+
+  return result_table;
+}
+
+std::shared_ptr<AbstractOperator> AggregateSort::_on_deep_copy(
+    const std::shared_ptr<AbstractOperator>& copied_input_left,
+    const std::shared_ptr<AbstractOperator>& copied_input_right) const {
+  return std::make_shared<AggregateSort>(copied_input_left, _aggregates, _groupby_column_ids);
+}
+
+void AggregateSort::_on_set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters) {}
+
+void AggregateSort::_on_cleanup() {}
+
+template <typename ColumnType>
+void AggregateSort::_write_aggregate_output(boost::hana::basic_type<ColumnType> type, ColumnID column_index,
+                                            AggregateFunction function) {
+  switch (function) {
+    case AggregateFunction::Min:
+      write_aggregate_output<ColumnType, AggregateFunction::Min>(column_index);
+      break;
+    case AggregateFunction::Max:
+      write_aggregate_output<ColumnType, AggregateFunction::Max>(column_index);
+      break;
+    case AggregateFunction::Sum:
+      write_aggregate_output<ColumnType, AggregateFunction::Sum>(column_index);
+      break;
+    case AggregateFunction::Avg:
+      write_aggregate_output<ColumnType, AggregateFunction::Avg>(column_index);
+      break;
+    case AggregateFunction::Count:
+      write_aggregate_output<ColumnType, AggregateFunction::Count>(column_index);
+      break;
+    case AggregateFunction::CountDistinct:
+      write_aggregate_output<ColumnType, AggregateFunction::CountDistinct>(column_index);
+      break;
+  }
+}
+
+template <typename ColumnType, AggregateFunction function>
+void AggregateSort::write_aggregate_output(ColumnID column_index) {
```

Comment:
Isn't this something that can be shared by the aggregate implementations?

Reply:
Yes, probably. Iirc we wrote/copied this function in two steps, and probably forgot to check if there are differences after the second version.

Reply:
Just checked the [hash aggregate implementation](https://github.com/hyrise/hyrise/blob/master/src/lib/operators/aggregate.cpp). There is a function with the same name, and mostly does the same, but there is a hash-specific part (L. 788-802). Further, it writes the aggregate outputs (L. 810-818), which we already do at the end of `_aggregate_values`.

I do not intend saying this cannot be moved up, its just not completely trivial

Reply:
In that case, please add a comment on the semi-duplication of the code and the reason why merging it does not improve it